### PR TITLE
[windows] add agent check for detecting windows BSOD

### PR DESCRIPTION
**Pull request by derekwbrown**
Adds a new check for sending an agent event if a blue screen is detected on agent start.

- Actual check is implemented in system probe because admin rights are needed to read crashdump

- Check records most recently recorded dump in registry, so event is only ever sent once

- Check itself runs only once; after first successful run it returns immediately

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: cmd/agent/subcommands/run/dependent_services_windows.go
Comment (brycekahle) [inline cmd/agent/subcommands/run/dependent_services_windows.go:58]: IMHO we don't need to keep things under `system_probe_config`, but you could do something like `windows_crash_check.enabled` instead

File: cmd/system-probe/config/config.go
Comment (brycekahle) [inline cmd/system-probe/config/config.go:162]: Why are there two enable config keys?

File: cmd/system-probe/modules/crashdetect_windows.go
Comment (brycekahle) [inline cmd/system-probe/modules/crashdetect_windows.go:23]: Can we make the naming consistent? I see "windows crash check", "windows crash detection", and "BSOD probe".

File: cmd/system-probe/config/config.go
Comment (derekwbrown) [reply cmd/system-probe/config/config.go:162]: good catch.  We moved the config around.  that should be removed.

File: cmd/system-probe/utils/utils.go
Comment (brycekahle) [inline cmd/system-probe/utils/utils.go:23]: leftover debug?

File: cmd/agent/subcommands/run/dependent_services_windows.go
Comment (derekwbrown) [reply cmd/agent/subcommands/run/dependent_services_windows.go:58]: the thinking was there's not additional config, so having an entire section seems excessive.  And, the OOM kill was enabled this way so it's consistent.

File: cmd/system-probe/utils/utils.go
Comment (derekwbrown) [reply cmd/system-probe/utils/utils.go:23]: yes.  🙇

File: cmd/agent/subcommands/run/dependent_services_windows.go
Comment (brycekahle) [reply cmd/agent/subcommands/run/dependent_services_windows.go:58]: There is no additional config...now 😎

Comment (brycekahle) [reply cmd/agent/subcommands/run/dependent_services_windows.go:58]: I'd consider the OOM kill way as legacy. All the other new products have their own namespaces, including the upcoming `ebpf` check.

**Review by drichards-87 — APPROVED** [review]

File: pkg/process/net/check_windows.go
Comment (hmahmood) [inline pkg/process/net/check_windows.go:27]: Should we bail early if `module != sysconfig.WindowsCrashDetectModule`?

Comment (derekwbrown) [reply pkg/process/net/check_windows.go:27]: this is taken from the equivalent in `check.go`.  

we could do an early return, but adding additional remote checks (as in linux) would result in having a big `if/else` both before and after the request.

I'd be happy to change it if you think it's necessary @hmahmood

Comment (hmahmood) [reply pkg/process/net/check_windows.go:27]: Running the check only to discard it later doesn't seem right. @xornivore could you shed some light on this please?

Comment (derekwbrown) [reply pkg/process/net/check_windows.go:27]: fair enough.  I'll change it but note the linux side then should have the same change.

**Review by hmahmood — APPROVED** [review]
For the NPM owned bits

File: pkg/config/config_template.yaml
Comment (sgnn7) [inline pkg/config/config_template.yaml:3916]: Could we add a bit more info here around this feature or a link to the public docs?

File: pkg/config/config_template.yaml
Comment (sgnn7) [inline pkg/config/config_template.yaml:3918]:
```suggestion
  # enabled: false
```

**Review by sgnn7 — APPROVED** [review]
Couple of nit comments on the ASC-owned files. I think we will also will need some CODEOWNERS updates included for the new files.

File: cmd/agent/subcommands/run/dependent_services_windows.go
Comment (brycekahle) [inline cmd/agent/subcommands/run/dependent_services_windows.go:58]: I think this is the correct path now, right?
```suggestion
			"windows_crash_detection.enabled": config.SystemProbe,
```

File: pkg/metadata/inventories/inventories.go
Comment (brycekahle) [inline pkg/metadata/inventories/inventories.go:477]:
```suggestion
	SetAgentMetadata(AgentSPWinCrashEnabled, config.SystemProbe.GetBool("windows_crash_detection.enabled"))
```

File: pkg/metadata/inventories/inventories.go
Comment (brycekahle) [inline pkg/metadata/inventories/inventories.go:113]: nit: Use full name `windows_crash_detection` or the module name `wincrashdetect` here

File: pkg/process/net/check.go
Comment (brycekahle) [inline pkg/process/net/check.go:59]: This file is for linux only, so does it need this?

File: pkg/process/net/check_windows.go
Comment (brycekahle) [reply pkg/process/net/check_windows.go:37]: Linux does already return an error if you use an invalid check name

**Review by brycekahle — CHANGES_REQUESTED** [review]
Small changes needed, mostly around config namespace

File: cmd/agent/subcommands/run/dependent_services_windows.go
Comment (derekwbrown) [reply cmd/agent/subcommands/run/dependent_services_windows.go:58]: yes

File: cmd/agent/subcommands/run/dependent_services_windows.go
Comment (brycekahle) [inline cmd/agent/subcommands/run/dependent_services_windows.go:58]: oops, looks like a `gofmt is needed`

File: pkg/config/config_template.yaml
Comment (brycekahle) [inline pkg/config/config_template.yaml:3915]:
```suggestion
  ## @param enabled - boolean - optional - default: false
```

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:117]: I wonder if there are other MS drivers before real offender. E,g. could be ddnpm.sys passing information to tcp.sys (or whatever) which crashed because we screwed up some passed memory

File: pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go
Comment (iglendd) [reply pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go:31]: I wonder if we should collect few more bits

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:126]: Would not be simpler to save a timestamp and collect dump/dumps information which happen after the time stamp to avoid loading and analyzing the same dump.

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:92]: I wonder if there is a more elegant mechanism to disable it in the scheduler

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:56]: How is DLL is loaded?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:46]: should we delete the comment?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:2]: What is _cgo_export.h for?

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:148]: Are vars typically specified in the beginning of a module?

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (iglendd) [inline pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:199]: Is second and subsequent lines will have extra space and it is desired?

File: pkg/config/config_template.yaml
Comment (iglendd) [reply pkg/config/config_template.yaml:3927]: I feel that the product is still being defined 🤷‍♂️

File: pkg/collector/corechecks/system/wincrashdetect/empty.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/empty.go:1]: because otherwise the linux build fails.

File: pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:14]: it requires that because the function signature, to be shared code, takes an `ebpf.Config` structure pointer

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:2]: in includes the forward declarations that cgo generates for functions implemented in `go`, specifically the `logLineCallback`

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:199]: yes.

**Review by just-chillin — APPROVED** [review]

File: cmd/system-probe/modules/crashdetect_windows.go
Comment (derekwbrown) [reply cmd/system-probe/modules/crashdetect_windows.go]: it was `blue screen probe`.  Changed to `cp`, for crash probe, which follows the convention that it's variable of type WinCrashProbe. I supose it could be `wcp`

File: pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go:41]: it could be BugCheckCode.

File: pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/win_crash_types.go:20]: updated with comment

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc]: it was removed.

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go]: moved to beginning

**Review by KSerrania — COMMENTED** [review]
Deferring to the Windows Kernel Integrations team, since all the files currently "owned" by Agent Platform in this PR are files that will be owned by that team (because of the `CODEOWNERS` change). Will approve once this gets an approval from this team.

**Review by paulcacheux — APPROVED** [review]

File: pkg/metadata/inventories/inventories.go
Comment (derekwbrown) [reply pkg/metadata/inventories/inventories.go:477]: @derekwbrown

File: pkg/config/config_template.yaml
Comment (derekwbrown) [reply pkg/config/config_template.yaml:3927]: The product is not still being defined.  This is the result of a PM-defined feature, with a product brief, and the output has been defined & approved by PM.

File: pkg/metadata/inventories/inventories.go
Comment (derekwbrown) [reply pkg/metadata/inventories/inventories.go:477]: good catch 🙇

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:55]: > How is DLL is loaded?

it is statically linked.  the `cgo` params in crashparams.go

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:79]: see newly attached readme

File: pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:45]: Well, it decodes the crash dump file name based on the currently configured settings.

File: pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:70]: This part of the code is simply for doing the admin-required work of checking whether there is a dump.  It is planned to be used by more than one consumer.

The current consumer, in `wincrashdetect.go`. does the work to make sure that the same crash isn't reported more than once.  And there are tests to verify.

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:100]: there is not.

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:134]: again, this is designed to be used by multiple consumers.

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (iglendd) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:55]: I see it now, thanks

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (iglendd) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:117]: I agree and we will see later if better heuristic could is possible

File: pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go
Comment (iglendd) [reply pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:45]: Yes, it is using settings only as assistance tool to get existing file path. Anyway it is only minor arguably more accurate semantically speaking note. Just a note if you would agree. It is ok otherwise as well.

File: pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go
Comment (iglendd) [reply pkg/collector/corechecks/system/wincrashdetect/wincrashdetect.go:134]: Not sure I understand what multiple consumers you are referring. End even if there are multiple consumers single timestamp bookmark would work as well. I it is just a suggestion though.

File: pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go
Comment (iglendd) [reply pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:70]: > The current consumer, in wincrashdetect.go. does the work to make sure that the same crash isn't reported more than once. And there are tests to verify.

Yes I have seen that code

>This part of the code is simply for doing the admin-required work of checking whether there is a dump. It is planned to be used by more than one consumer.

I think I am talking about other potential issue.  E.g. if you install Agent we expecting to pickup a crashdump happens before installation, right? Now imagine that there is a full memory dump but machine settings had been modified to report minidump after full crash dump generated (or other way around). Which means you will miss previous crash dumps. Of course it is a bit abstract and probably rare situation. The other "problem" that you will miss previous minidumps if there are more than one.

**Review by iglendd — APPROVED** [review]
looks good

File: pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/wincrashprobe.go:70]: yes.  But both of these are nominally intentional.
1) if there are crash dump(s) around from prior to the agent being installed, and they've been messing with the config, we will only see the most recent of the type currently configured.  It's not quite intentional, but it is an acceptable artifact of this implementation.  We're really only looking for things that happen after we're installed.

2) Yes, we will miss previous minidumps.  That's intentional.  The intent is to only report one, the most recent.

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:55]: I think we should check the return value of `DebugCreate` prior to dereferencing `g_Client`.

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashdump.cc:56]: I think we need to double check which of these functions have a return value, and check it.

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:29]: It isn't obvious what `maxscan` is - could you update the name, add a comment, or both? Also, for `200`, what is this representing and why did you choose it?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:31]: I think there's a `min` function in the Go standard lib, can you use that instead?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:68]: Is there a reason "RetAddr" isn't included in the `const()` section with "Bugcheck code" and "Debug session time"? If no, can it be?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:101]: This comment doesn't seem to apply to the subsequent line, does it belong here or on `110`?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (richl-dd) [inline pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:118]: I don't think this check is needed (at least, it isn't obvious to me why). If it is, could you explain? If it isn't, I think we can remove it.

Also, am I correct in assuming that this loop iterates the output (from your readme) and the things it cares to pull out include:
  - The time the debug session was kicked off
  - The bugcheck prefix
  - The first callsite that isn't `nt!`

?

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:31]: done.

File: pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go
Comment (derekwbrown) [reply pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go:118]: I added some sample lines in comment.  But there could be "unable to load image" interlaced with lines we want, in which case the search for `NT!` would fail but the line isn't the one we want.  Hopefully it's more clear with the sample lines comment to be added.

File: cmd/agent/dist/conf.d/wincrashdetect.d/conf.yaml.example
Comment (sgnn7) [inline cmd/agent/dist/conf.d/wincrashdetect.d/conf.yaml.example:5]:
```suggestion
    ## The Windows crash detection check will send an event if a system crash is detected
```

**Review by sgnn7 — APPROVED** [review]
Couple of more comments but all non-blocking nits

**Review by brycekahle — APPROVED** [review]
Looks OK from ebpf-platform view, but it looks like some legitimate issues have been caught by the unit-tests/linter.

**Review by ISauve — APPROVED** [review]

**Pull request merged**
